rails/empty-arrays-in-json-request.md

```diff
@@ -0,0 +1,66 @@
+Deep Munging
+
+Even when turned off
+
+```ruby
+put "/api/v1/post/1", {
+  post: { tags: [] },
+  format: :json
+}
+```
+
+
+Will remove the entire `post` hash:
+
+```ruby
+{}
+```
+
+Since the `"Content-Type"` is unset, Rack defaults to treat the request as
+`"application/x-www-form-urlencoded"`.
+
+The URL encoded representation of `["foo", "bar"]` is:
+
+```
+?tags[]=foo&tags[]=bar
+```
+
+However, it isn't possible to represent `[]`:
+
+```
+?tags[]=
+```
+
+As a result, Rack ignores the `tags` key, which would make the `post` key point
+to an empty hash, which Rack also ignores.
+
+To work around this issue:
+
+* transforming the `params` hash to a [`JSON` string][rack-test]
+* appending `.json` to the URL
+* explicitly set the `"Content-Type"` head to `"application/json"`
+
+```ruby
+put "/api/v1/post/1.json", {
+  post: { tags: [] }
+}.to_json, { "Content-Type" => "application.json" }
+```
+
+This pattern can be extracted to a helper:
+
+```ruby
+def json_put(path, params, headers_or_env = {})
+  json_path = "#{path}.json"
+  params_without_format = params.except(:format)
+  headers = headers_or_env.reverse_merge("Content-Type" => "application/json")
+
+  put(json_path, params_without_format, headers)
```

Comment:
Yeah, you'd need to call `.to_json`. It doesn't matter if something is a query param, or part of the request body.